cilium/cmd/cleanup.go

```diff
@@ -0,0 +1,242 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/daemon/defaults"
+	"github.com/cilium/cilium/pkg/bpf"
+
+	"github.com/spf13/cobra"
+	"github.com/vishvananda/netlink"
+)
+
+// configCmd represents the config command
+var cleanupCmd = &cobra.Command{
+	Use:   "cleanup",
+	Short: "Reset the agent state",
+	Run: func(cmd *cobra.Command, args []string) {
+		common.RequireRootPrivilege("cleanup")
+		runCleanup()
+	},
+}
+
+var force bool
+
+const (
+	ciliumLinkPrefix = "cilium_"
+	hostLinkPrefix   = "lxc"
+	hostLinkLen      = len(hostLinkPrefix + "XXXXX")
+	tunnelMap        = "tunnel_endpoint_map"
+	cniConfig        = "/etc/cni/net.d/10-cilium-cni.conf"
+)
+
+func init() {
+	rootCmd.AddCommand(cleanupCmd)
+	cleanupCmd.Flags().BoolVarP(&force, "force", "f", false, "Skip confirmation")
+}
+
+func runCleanup() {
+	// Abort if the agent is running, err == nil is handled correctly by Stat.
+	if _, err := os.Stat(defaults.PidFilePath); !os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "Agent should not be running when cleaning up\n"+
+			"Found pidfile %s\n", defaults.PidFilePath)
+		os.Exit(1)
+	}
+
+	routes, links, err := findRoutesAndLinks()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+	}
+
+	showWhatWillBeRemoved(routes, links)
+	if !force {
+		if !confirmCleanup() {
+			return
+		}
+	}
+
+	// ENOENT and similar errors are ignored. Should print all other
+	// errors seen, but continue.  So that one remove function does not
+	// prevent the remaining from running.
+	type cleanupFunc func() error
+	checks := []cleanupFunc{removeAllMaps, unmountFS, removeDirs, removeCNI}
+	for _, clean := range checks {
+		if err := clean(); err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+		}
+	}
+
+	if err := removeRoutesAndLinks(routes, links); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+	}
+}
+
+func showWhatWillBeRemoved(routes map[int]netlink.Route, links map[int]netlink.Link) {
+	fmt.Printf("Warning: Destructive operation. You are about to remove:\n"+
+		"- all BPF maps in %s containing '%s' and '%s'\n"+
+		"- mounted bpffs at %s\n"+
+		"- library code in %s\n"+
+		"- endpoint state in %s\n"+
+		"- CNI configuration at %s\n",
+		bpf.MapPrefixPath(), ciliumLinkPrefix, tunnelMap, bpf.GetMapRoot(),
+		defaults.LibraryPath, defaults.RuntimePath, cniConfig)
+
+	if len(routes) > 0 {
+		fmt.Printf("- routes\n")
+		for _, v := range routes {
+			fmt.Printf("%v\n", v)
```

Comment:
Is there any chance we can make this printout a little more human readable? The route object was ok but the links one was difficult to understand because it's full of nils. Unfortunately, there doesn't seem to be a nice String() function on any of these objects... I tried "%+v" and it printed the field names, at least, which make it easier to read.
```
- routes
{Ifindex: 187 Dst: 169.254.254.0/24 Src: <nil> Gw: <nil> Flags: [] Table: 254}
- links
&{LinkAttrs:{Index:11 MTU:1450 TxQLen:0 Name:lxc15cdb HardwareAddr:5e:8d:e4:9d:64:02 Flags:up|broadcast|multicast RawFlags:69699 ParentIndex:10 MasterIndex:0 Namespace:<nil> Alias: Statistics:0xc4215066d0 Promisc:0 Xdp:<nil> EncapType:ether Protinfo:<nil> OperState:up} PeerName:}
&{LinkAttrs:{Index:183 MTU:1450 TxQLen:0 Name:lxc41b3f HardwareAddr:ea:57:7f:12:87:05 Flags:up|broadcast|multicast RawFlags:69699 ParentIndex:182 MasterIndex:0 Namespace:<nil> Alias: Statistics:0xc421506bc8 Promisc:0 Xdp:<nil> EncapType:ether Protinfo:<nil> OperState:up} PeerName:}
&{LinkAttrs:{Index:186 MTU:1500 TxQLen:1000 Name:cilium_net HardwareAddr:56:20:e9:9c:b5:5b Flags:up|broadcast|multicast RawFlags:69827 ParentIndex:187 MasterIndex:0 Namespace:<nil> Alias: Statistics:0xc4215070ec Promisc:0 Xdp:<nil> EncapType:ether Protinfo:<nil> OperState:up} PeerName:}
&{LinkAttrs:{Index:187 MTU:1500 TxQLen:1000 Name:cilium_host HardwareAddr:7a:0a:65:7e:d5:e9 Flags:up|broadcast|multicast RawFlags:69827 ParentIndex:186 MasterIndex:0 Namespace:<nil> Alias: Statistics:0xc4215075dc Promisc:0 Xdp:<nil> EncapType:ether Protinfo:<nil> OperState:up} PeerName:}
```

Reply:
Let's leave this as is and if it gets too much we can pretty print later.

Reply:
Please file a microtask GH issue to write nice output functions. It's a great entry level task.

Reply:
Issue created https://github.com/cilium/cilium/issues/2104